### ISSUE TITLE
Folding code: more advanced commands

### ISSUE THREAD

**Issue by ghost** [geany/geany#2920]
EDIT: updated some of suggested commands: fold/ unfold of descendant nodes of given node.

[Currently](https://www.geany.org/manual/current/index.html#document-keybindings), the only ways to change the folding is:
- toggle current fold
- fold all
- unfold all

For files with many nodes and deep levels of children nodes, it would be great to have commands for efficient navigation and (controlled) fold/unfold of the hierarchical structure: 
* **CO1** Move from current node into the parent node (if it has one) and fold that
* **CO2** the opposite: Unfold current node (if parent) and move to 1st child node.

When applied repeatedly, will allows to browse and fold/unfold a full branch / path of nodes. Similar to navigation with arrows in   many file managers' tree view of folders (right/left to move across levels and collapse/expand, and up/down to jump  through sibling nodes)

* **C3** Folding of all ~~parent nodes in the file at same level as current node's level~~ descendant nodes of current node. 
* **C4** Unfolding of all ~~parent nodes in the file at same level as current node's level~~ descendant nodes of current node. 

This will fill the big control gap between what's available currently: folding at current node and at whole document. They are analogous to how current "fold all" and "unfold all" work, if think of the file as the "root node" of all other nodes/lines. 
These would be more useful, IMO, than the fold/unfold of all nodes  of same fixed-level that some other editors provide (for ex , Notepad++ ). First, it allows local fold-region control (so that does change folding in distand sibling nodes/branches, and don't load your visual memory with too much stuff). Second, it's relative to current node, so you don't not need to know which level exactly you're currently on.

By _node_ I don't mean only something that can be folded, but any "logical line" (non-blank line; with one or more programming statements on it; perhaps comment lines too, depending on language ).  
A node is a  parent node if it contains a folding point (and the nodes inside the fold - children nodes); otherwise is a leaf node. It's a child node if it's inside the fold of another node.
```python
# comment                 # level 0, leaf
print(9)                  # level 0, leaf node
    # comment             # level 0  leaf node (b/c previous line is not a parent node)
def f(x):                 # level 0, parent node
    """ doc string        # level 1, parent node
    lalala                # level 2, leaf node
    """
    y=x+1                 # level 1 , leaf node
    for i in L:           # level 1, parent
        do(x)             # level 2, leaf
    print(x-y)            # level 1, leaf node
    # comment             # level 1, leaf
# comment                 # level 0, leaf node
print(3)                  # level 0, leaf node
```
In my own use case, in addition to coding, I use Geany for organizing notes in `.txt` files, with folding by indentation. With ability to navigate fast, and control the hierarchy display, I will not need to use a separate "outliner".

**Comment by elextr** [comment]
I don't use folding much myself, but this looks reasonable, perhaps fold users could comment.

But a couple of points, the keybindings should have menu equivalents so they are more discoverable or the actions usable if the keys are not bound.  I don't like using up additional keys for default bindings, its already getting so there are not many left, so they would be unbound by default and the menu would allow occasional use of the functionality without binding them.  Users with a high usage would of course bind keys to the actions.  

I would suggest that in the `Document` menu `Fold All` and `Unfold All` with a single `Fold...` and a submenu with all the 9 commands.

**Comment by VictorelPetrovich** [comment]
Thanks for commenting; I understand.  
My last edit about meaning of "node" was not suggest placing "fold points" at other places than defined by lexers, but to clarify how the navigation might proceed for "nodes at same level". But if it will be too hard to treat lines with fold points and without as being still nodes at same level, then I suppose restricting "nodes" only to fold points can still be useful.

**Comment by VictorelPetrovich** [comment]
Of course, folding can only happen at a fold points. 
But what is considered as "node" in the tree of the file could be made to be more inclusive than the fold points. Because that changes the determination of the current level and thus the navigation and which fold points will be folded/unfolded. 
For example, in the text my previous comment:
- if currently the cursor is in middle of line 2, and one uses C3 to go to next sibling, I think that should be line 4 (as opposed to skiping it b/c it's not a fold point)
- if currently in line 8, and one uses C5 (to fold all nodes of same level as current node), then IMO it should fold all folding points of level 1 - judging by current indentation level (as opposed to saying that current fold's level is 0, thus fold all folding points of level 0)

**Comment by VictorelPetrovich** [comment]
To be clearer, for command C1: if the "node" is a leaf, then only fold if it actually has children (is a fold point, per current lexer definition).

**Comment by VictorelPetrovich** [comment]
Here's a first draft for CO1, based on https://www.scintilla.org/ScintillaDoc.html, especially https://www.scintilla.org/ScintillaDoc.html#Folding*/.
**But** I'm a beginner in C, I've never contributed code to online projects before, and I have no idea where exactly this would be integrated in Geany. 
So any advice/info is welcome to make this more fun :) 
``` c
#include "ScintillaTypes.h"
#include "ScintillaMessages.h"

/* move cursor to parent line (if current line has a parent) and fold it (hide its children lines). */
void moveToParentAndFold(line currentLine)
{
	line parentLine= SCI_GETFOLDPARENT(currentLine);
	// if currently at level 0 (no parents), then effect will be folding current line
	if (parentLine == -1)
		parentLine=currentLine;

	//set caret at first non-whitespace char in parentLine
	SCI_GOTOPOS(SCI_GETLINEINDENTPOSITION(parentLine) );

	// fold parentLine
	SCI_FOLDLINE(parentLine, SC_FOLDACTION_CONTRACT);
}
```

**Comment by VictorelPetrovich** [comment]
Here's a first draft for CO1, based on https://www.scintilla.org/ScintillaDoc.html, especially https://www.scintilla.org/ScintillaDoc.html#Folding.
**But** I'm a beginner in C, I've never contributed code to online projects before, and I have no idea where exactly this would be integrated in Geany. 
So any advice/info is welcome to make this more fun :) 
``` c
// for `line` type, https://www.scintilla.org/ScintillaDoc.html
#include "ScintillaTypes.h"
#include "ScintillaMessages.h"


/* move cursor to parent line (if current line has a parent) and fold it (hide its children lines).
 * if currently at level 0 (no parents), then effect will be folding current line
 */
void moveToParentAndFold()
{
	currentLine = SCI_LINEFROMPOSITION(SCI_GETCURRENTPOS);

	line parentLine= SCI_GETFOLDPARENT(currentLine);
	if (parentLine == -1)
		parentLine=currentLine;

	//set caret at first non-whitespace char in parentLine
	SCI_GOTOPOS(SCI_GETLINEINDENTPOSITION(parentLine) );

	SCI_FOLDLINE(parentLine, SC_FOLDACTION_CONTRACT);

}
```

**Comment by VictorelPetrovich** [comment]
Since the primary motivation for folding is to prevent visual information overload, I think there's not much value in expanding nodes, at same level, across _all trees _ in the entire file. (By the way, for the same reason, I'd choose to have all my files open by default with all nodes folded. )
So that's why I made the changes above to C3 and C4, to be able to fold/unfold all nodes _locally_, belonging to same parent node/subtree (As complement to fold/unfold at whole file level, and individual node level).

@intact , what did you have in mind with Fold Recursively, Unfold Recursively ?

**Comment by VictorelPetrovich** [comment]
> Btw looks like "Toggle current fold" use that "Fold/unfold all children of a fold point" configuration too.

Right; except the Shift switch doesn't work to change the behavior of the Toggle current command.

So then your (un)fold recursive is almost the same as what I got to in C3/C4. With just one difference: to apply to descendants (= to children, recursively) only, not to parent node.
Here's why it may be better: 
when you come to a parent node:
1. if it's initially closed, your first action will probably be just to toggle it open. You wouldn't want to open it and its children recursively, as that will reveal too much infor at once. 
2. Now once you opened it, if children happen to be closed, that's fine, you can select one of them to explore (in particular with CO1/CO2 if want). If you really want to pop them all open at once, then the Unfold All descendants I suggest still does the job. 
But if all/most chilren are already opened (let alone if their descendants too), you'll want to close them (too much info at once!). So here the Fold All Descendants will come in very handy: will fold the children of children, but you'll still see curent (parent) and its children.
3. If instead the node is initially open... see 2.

(This formulation also logically agrees with what (Un)Fold All does, if imagine that the ancestor node  of all nodes in file is the file itself..)

So we have 3 existing, + 4 new suggested. 
In addition, we could also include a command to open/unfold All nodes of Ist (top) level, and another to close/fold them   - just b/c we're at it... 
Though I'm personally not convinced these fixed-level of all nodes are useful; or at least not useful beyond Ist level (b/c usually you dont' want to see too much at once, and you'll work in a local region anyway, where toggle current and/or (Un)Fold Descendants and CO1/CO2 will do a better job ).

**Comment by PthDE** [comment]
I'm a simple hobbyist, enjoying Geany as my main editor when coding for use under the Arduino IDE.
When both applications run simultaneously, Geany correctly detects an opened file being changed by another application, for example due to edits made and saved via the Arduino IDE.
Re-loading the file in Geany causes all code-folding that was previously set, to be lost. It's a small irritation given the many features Geany that offers, but of course "another nice to have".
Many thanks

**Comment by elextr** [comment]
@PthDE IIRC one of the plugins might remember folding, but how good (or not) it is if the folds have changed I don't know.
